jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/RabitTracker.scala

```diff
@@ -0,0 +1,193 @@
+
+import java.net.{InetAddress, InetSocketAddress}
+
+import akka.actor.ActorSystem
+import akka.pattern.ask
+import ml.dmlc.xgboost4j.java.IRabitTracker
+import ml.dmlc.xgboost4j.scala.rabit.handler.RabitTrackerHandler
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Scala implementation of the Rabit tracker interface without Python dependency.
+  * The Scala Rabit tracker fully implements the timeout logic, effectively preventing the tracker
+  * (and thus any distributed tasks) to hang indefinitely due to network issues or worker node
+  * failures.
+  *
+  * Note that this implementation is currently experimental, and should be used at your own risk.
+  *
+  * Example usage:
+  * {{{
+  *   import scala.concurrent.duration._
+  *
+  *   val tracker = new RabitTracker(32)
+  *   // allow up to 10 minutes for all workers to connect to the tracker.
+  *   tracker.start(10 minutes)
+  *
+  *   /* ...
+  *      launching workers in parallel
+  *      ...
+  *   */
+  *
+  *   // wait for worker execution up to 6 hours.
+  *   // providing a finite timeout prevents a long-running task from hanging forever in
+  *   // catastrophic events, like the loss of an executor during model training.
+  *   tracker.waitFor(6 hours)
+  * }}}
+  *
+  * @param numWorkers Number of distributed workers from which the tracker expects connections.
+  * @param port The minimum port number that the tracker binds to.
+  *             If port is omitted, or given as None, a random ephemeral port is chosen at runtime.
+  * @param maxPortTrials The maximum number of trials of socket binding, by sequentially
+  *                      increasing the port number.
+  * @param tcpBindTimeout Timeout for the tracker to bind to an InetSocketAddress.
+  */
+class RabitTracker private[scala](numWorkers: Int, port: Option[Int] = None,
+                                  maxPortTrials: Int = 1000, tcpBindTimeout: Duration = 1 minute)
+  extends IRabitTracker {
+
+  import scala.collection.JavaConverters._
+
+  require(numWorkers >=1, "numWorkers must be greater than or equal to one (1).")
+
+  val system = ActorSystem.create("RabitTracker")
+  val handler = system.actorOf(RabitTrackerHandler.props(numWorkers), "Handler")
+  implicit val askTimeout: akka.util.Timeout = akka.util.Timeout(30 seconds)
+
+  val futureWorkerEnvs: Future[Map[String, String]] = Try(
+    Await.result(handler ? RabitTrackerHandler.RequestBoundFuture, askTimeout.duration)
+      .asInstanceOf[Future[Map[String, String]]]
+  ) match {
+    case Success(fut) => fut
+    case Failure(e) =>
+      val delayedFailure = Promise[Map[String, String]]()
+      delayedFailure.failure(e)
+      delayedFailure.future
+  }
+
+  // a future for XGBoost worker execution
+  val futureCompleted: Future[Int] = Try(
+    Await.result(handler ? RabitTrackerHandler.RequestCompletionFuture, askTimeout.duration)
```

Comment:
The future here is just to mirror a `Promise` object in the Akka actor, because training completion will occur in the future when the Promise is fulfilled. 

It has a `Await.result` here because ask pattern always return a Future. Basically it's a `Future[Future[Int]]` type, which I need to unpack to get the actual `Future`, which `waitFor` will wait for.